src/csharp/Grpc.Core/Internal/AsyncCall.cs

```diff
@@ -648,13 +648,13 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
             if (status.StatusCode != StatusCode.OK)
             {
                 streamingResponseCallFinishedTcs.SetException(new RpcException(status, receivedStatus.Trailers));
-                if (status.StatusCode == StatusCode.Cancelled || origCancelRequested)
+                if (status.StatusCode == StatusCode.Cancelled || status.StatusCode == StatusCode.Unavailable || origCancelRequested)
                 {
                     // Make sure the exception set to the Task is observed,
```

Comment:
For extra context, here's my understanding of why we're seeing the UnobservedException at all. 
I think the issue is with the way the streaming responses are processed.
The call.ResponseStream.MoveNext() reads the next response message and only once it  gets reaches a null response,
it observes the result of  `call.StreamingResponseCallFinishedTask` here:
https://github.com/grpc/grpc/blob/1cd6e69347cbf62a012477fe184ee6fa8f25d32c/src/csharp/Grpc.Core/Internal/ClientResponseStream.cs#L59

So if MoveNext() is not running (or if it doesn't exhaust the response stream) the exception coming from call.StreamingResponseCallFinishedTask will never be observed (which later shows up in finalizer as unobserved exception).

Note that the call.StreamingResponseCallFinishedTask "starts" as soon a the call is started (but there may or may not be a call to responseStream.MoveNext().